site/content/blog/2020-11-17-mysql-index-merge.md

```diff
@@ -0,0 +1,178 @@
+  `IsPinned` tinyint(1) DEFAULT NULL,
+  `UserId` varchar(26) DEFAULT NULL,
+  `ChannelId` varchar(26) DEFAULT NULL,
+  `RootId` varchar(26) DEFAULT NULL,
+  `Message` text,
+  PRIMARY KEY (`Id`),
+  KEY `idx_posts_update_at` (`UpdateAt`),
+  KEY `idx_posts_create_at` (`CreateAt`),
+  KEY `idx_posts_delete_at` (`DeleteAt`),
+  KEY `idx_posts_channel_id` (`ChannelId`),
+  KEY `idx_posts_root_id` (`RootId`),
+  KEY `idx_posts_user_id` (`UserId`),
+  KEY `idx_posts_channel_id_update_at` (`ChannelId`,`UpdateAt`),
+  KEY `idx_posts_channel_id_delete_at_create_at` (`ChannelId`,`DeleteAt`,`CreateAt`),
+  FULLTEXT KEY `idx_posts_message_txt` (`Message`),
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4
+```
+
+This is an abridged version of the `Posts` table for brevity. But it contains the essential elements for us to understand the problem. Our main columns of interest are `CreateAt`, `DeleteAt`, `UpdateAt`, and `ChannelId`. Each of them have their individual indices, and there are 2 additional multi-column indices. One having `ChannelId` and `UpdateAt`. Another involving `ChannelId`, `DeleteAt`, and `CreateAt`.
+
+Now let's look into the query again.
+
+```sql
+SELECT Id FROM Posts WHERE ChannelId = 'x' AND DeleteAt = y AND CreateAt < z
+	ORDER BY CreateAt DESC
+	LIMIT 1
+```
+
+We are filtering the table with 3 columns- `ChannelId`, `DeleteAt`, and `CreateAt`. And then ordering by `CreateAt` and just getting the first row. It becomes clear why choosing the multi-column index is faster than choosing the one for `CreateAt`. Simply because the query filters with all 3 columns which leads to scanning a much smaller dataset. But then, why is MySQL choosing the wrong index in the first place? Using `USE INDEX` was the nuclear button which I didn't want to use unless there were no other options.
+
+I wasn't able to reproduce the problem locally, so it was not possible to try different variations of the query. But after some googling, I had a pretty good theory of what was happening. It was the "ORDER BY" clause. MySQL tries to be smart and decides that although using the `CreateAt` index might have to scan through more rows, it avoids the sorting at the end. Whereas actually, using the multi-column index leads to scanning fewer rows, which gets sorted in practically no time at all.
+
+Now that we understand the problem, how can we coax MySQL into choosing the right index? Well, if MySQL is acting smart, we can outsmart it. What could possibly go wrong? Right?
+
+Since the choice of the index becomes dictated with the "ORDER BY" clause rather than the WHERE clause, what if we can include that decision in the "ORDER BY" clause itself? If we change `ORDER BY CreateAt` to `ORDER BY ChannelId, DeleteAt, CreateAt`, the query result remains exactly the same. Because `ChannelId` and `DeleteAt` are equality checks. But now MySQL goes- "Aha, now I have to sort by these 3 columns. So I better use the multi-column index". And that's exactly what we want!
```

Comment:
```suggestion
Since the choice of the index is dictated by the "ORDER BY" clause rather than the "WHERE" clause, what if we can include that decision in the "ORDER BY" clause itself? If we change `ORDER BY CreateAt` to `ORDER BY ChannelId, DeleteAt, CreateAt`, the query result remains exactly the same. Because `ChannelId` and `DeleteAt` are equality checks. But now MySQL goes "Aha, now I have to sort by these three columns. So I'd better use the multi-column index". And that's exactly what we want!
```